CORDA-3381 Errors in vault updates publisher are unsubscribing stopping observers from working

kyriathar opened the pull request:
Up to now, whenever an `rx.Observer` would get subscribed (not unsafe subscribe) to an `rx.Observable` it would get wrapped with a `rx.SafeSubscriber`. If an error would occur in `Observer.onNext`, its SafeSubscriber (wrapping the Observer) was unsubscribing. Like so, the `rx.Observer` was no longer available the next time an event would be pushed from the `rx.Observable`, leading to malfunctioning CordApps.
 
With this PR, we will introduce a new type of `rx.Subscriber` (called `ResilientSubscriber`) which extends `rx.SafeSubscriber`, however upon catching an error from its Observer 
it will not set `SafeSubscriber.done` flag to true nor will call `SafeSubscriber.unsubscribe`. 
Furthermore, we will introduce a new `rx.OnSubscribe` (called `OnResilientSubscribe`) which
upon subscribe, will unwrap the `rx.Observer` from the `rx.SafeSubscriber`, re-wrap it with `ResilientSubscriber` and then subscribe it to the source `rx.Observable`.

`VaultService.rawUpdates` will exposed to `CordaService`s through an Observable with a `OnResilientSubscribe` to make subscribing Observers resilient to errors that will come out of `onNext`.
We will also provide `ResilientSubscriber`s to users if they wish to subscribe as such via a new public API method (`Observable.continueOnError`).

kyriathar-r3 committed:
Throw SQLException or PersistenceException plain, that may come out of an unsafe subscriber

lankydan commented on core/src/main/kotlin/net/corda/core/internal/InternalUtils.kt:
```diff
@@ -172,8 +173,8 @@ fun <T> Observable<T>.bufferUntilSubscribed(): Observable<T> {
     val subject = PublishSubject.create<T>()
-    subject.subscribe(this)
-    teeTo.forEach { subject.subscribe(it) }
+    subject.unsafeSubscribe(Subscribers.from(this))
```
Add a comment about why we call `unsafeSubscribe`. Basically just say that a failure in a single subscriber will cause a `PublishSubject` to unsubscribe all of its child subscribers

lankydan replied:
You need to rebase this branch onto 4.4 now because you already added this come in another PR

kyriathar replied:
Rebase done - skipped this commit since it was merged with PR #5913

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +190,30 @@ class ObservablesTests {
+     * eventually shut down all of the subscribers under that PublishSubjectState.
+     */
+    @Test
+    fun `error in unsafe subscriber won't shutdown subscribers under same publish subject, after tee`() {
```
good idea to have this test 👍

lankydan replied:
Same as before, rebase as I think you already merged this.

kyriathar replied:
same as before - rebase done - skipped this commit since it was merged with PR #5913

lankydan commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -414,7 +414,7 @@ class NodeVaultService(
                             }
                         }
-                    } ?: HospitalizeFlowException(e)
+                    } ?: (e as? SQLException ?: (e as? PersistenceException ?: HospitalizeFlowException(e)))
```
We can maybe tidy this up another time.

kyriathar-r3 committed:
Add explanatory comment about why we changed Observer.tee to use unsafe subscribe

kyriathar-r3 committed:
Introducing not unsubscribing version of Rx.Subscriber

kyriathar-r3 committed:
Wrap PublishSubjects with FlowSafeSubjects in all tests that test Observer.tee

kyriathar-r3 committed:
Minor code formatting

kyriathar-r3 committed:
Make rawUpdates Rx.Observers not unsubscribe when accessed from CordaServices - Do not allow rawUpdates subscribing from flows

kyriathar-r3 committed:
Warning fix: Add else block to when statement

kyriathar-r3 committed:
Revert "Wrap PublishSubjects with FlowSafeSubjects in all tests that test Observer.tee"

This reverts commit e419af86

kyriathar-r3 committed:
Correcting log message

kyriathar-r3 committed:
Improve log message

kyriathar-r3 committed:
Add fiber's id to log message and exception message

kyriathar-r3 committed:
Added test, asserting FlowSafeSubscriber is alive and re-accessed upon flow retry

lankydan commented on core/src/main/kotlin/net/corda/core/internal/FlowSafeSubscriber.kt:
```diff
@@ -0,0 +1,67 @@
+                // this Subscriber may wrap a non leaf Observer. In case the wrapped Observer is a PublishSubject then we
+                // should not call onError because PublishSubjectState.onError will shut down all of the Observers under it
+                throw OnNextFailedException(
+                    "Observer.onNext failed, this is a non leaf FlowSafeSubscriber, therefore onError will be skipped", e
```
Say this as well:

"We throw [OnNextFailedException] to pass the exception through the standard [Subscriber] chain of events without triggering any [onError]s or being caught by other [Subscriber]s"

lankydan commented on core/src/main/kotlin/net/corda/core/internal/FlowSafeSubscriber.kt:
```diff
@@ -0,0 +1,67 @@
+}
+
+@VisibleForTesting
+class OnNextFailedException(message: String, cause: Throwable) : OnErrorNotImplementedException(message, cause)
```
Add kdoc for this class. Maybe:

```
[OnNextFailedException] extends [OnErrorNotImplementedException] to allow it to pass through RxJava exception
handling without being caught by calls to [Exceptions.throwIfFatal] while better detailing why we are throwing the exception.

lankydan commented on core/src/main/kotlin/net/corda/core/utilities/FlowSafeSubject.kt:
```diff
@@ -0,0 +1,45 @@
+ *
+ * Preventing this is useful to observers that are subscribed to important observables to prevent them from ever unsubscribing due to an
+ * error. Unsubscribing could lead to a malfunctioning CorDapp, for the rest of the current run time, due to a single isolated error.
+ */
```
We should specifically talk about how this is to ensure that flows can experience errors and be handled by the state machine without unsubscribing the subscribers.

kyriathar replied:
"flows experiencing errors in rx.Observers but their wrapping rx.Subscribers are not getting unsubscribed" could perhaps better be mentioned in kdoc of FlowSafeSubscriber.

lankydan commented on core/src/main/kotlin/net/corda/core/utilities/FlowSafeSubject.kt:
```diff
@@ -0,0 +1,45 @@
+ * error. Unsubscribing could lead to a malfunctioning CorDapp, for the rest of the current run time, due to a single isolated error.
+ */
+@VisibleForTesting
+class FlowSafeSubject<T, R>(private val actual: Subject<T, R>) : Observer<T> by actual,
```
This should be in an internal package. Furthermore, we should be able to define this in `node` instead of `core`? I think it is only used by `NodeVaultService` so there is no reason to expose it further.

lankydan commented on core/src/main/kotlin/net/corda/core/utilities/FlowSafeSubject.kt:
```diff
@@ -0,0 +1,45 @@
+@VisibleForTesting
+class FlowSafeSubject<T, R>(private val actual: Subject<T, R>) : Observer<T> by actual,
+    Subject<T, R>(OnSubscribe<R> { subscriber ->
+        if (subscriber::class == SafeSubscriber::class) {
```
Add a comment to why we used `subscriber::class == SafeSubscriber::class` instead of `subscriber is SafeSubscriber`. A.k.a, that we will allow developers to write their own `SafeSubscriber`s and use them but we will replace all default implementations of `SafeSubscriber`.

Maybe even just move it out to a little single line method along with the comment (you dont have to do this part).  If we do this, it should be named something like `isDefaultSafeSubscriber` or `isNonSubclassedSafeSubscriber`, although both of those look a bit crap.

kyriathar replied:
Only because of the naming of that method adding confusion, maybe we would rather keep the expression inlined as is?

kyriathar-r3 committed:
Logging flow name instead of flow id at VaultService.rawUpdates subscribing error

lankydan commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -209,7 +210,32 @@ class NodeVaultService(
+                                "- the subscription did not succeed " +
+                                "- aborting the flow "
+                    )
+
```
Do we need this code + the `PreventSubscriptionsSubject`? Shouldn't logging and throwing the exception as soon as we call `get` be enough?

kyriathar replied:
Yes good point, i just thought because it is public API, then make least possible changes. In that sense, still return the rx.Observable but prevent from subscribing to it.

lankydan commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -209,7 +210,32 @@ class NodeVaultService(
+
+                    throw FlowException(
+                        "Flow ${it.logic::class.java.name} tried to subscribe an Rx.Observer to VaultService.rawUpdates " +
+                                "- Rx.Observables should only be subscribed outside the context of a flow " +
```
`- Rx.Observables should only be subscribed outside the context of a flow ` should be `- Rx.Observables should only be subscribed to outside the context of a flow `. The same applies for the log message above.

lankydan commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -209,7 +210,32 @@ class NodeVaultService(
+            }
+            // we are not inside a flow, we are most likely inside a CordaService;
+            // we will wrap with 'flow safe subscriptions' here;
+            // every Observable.subscribe to this object should be wrapped with a FlowSafeSubscriber (-not unsubscribe- subscribers).
```
Tidy up these comments. Mainly the last line

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+            source.onNext(2) // first observer will run, second observer will run and throw
+        }
+        source.onNext(3) // both observers will run
+        assertThat(heartBeat == 5)
```
Use `assertEquals`

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+            source.onNext(2) // first observer will run, second observer will run and throw
+        }
+        source.onNext(3) // both observers will run
+        assertThat(heartBeat == 5)
```
Is it correct that the `heartBeat` is 5? Shouldn't it increase by 2 for each event passed into `onNext`? Making it 6?

I think it would also be worth splitting out the `heartBeat` counts into multiple variables. So it is 100% clear how many times each subscriber is called (since there is only 2 of them)

kyriathar replied:
i think 5 is correct; 
run_1: the first Observer will increase counter and throw - the second will not be executed (+1)
run_2: first will increase counter - second will increase counter and throw (+2)
run_3: both Observers will run normally (+2)

`heartBeat` splitted into 2 variables

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+            }
+        })
+        source.subscribe { heartBeat += it }
+        // wrap FlowSafeSubject with a FlowSafeSubscriber
```
Is this comment correct?

kyriathar replied:
Valid point,  "FlowSafeSubscriber" should be "SafeSubscriber"

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+     * which will shut down all of the Subscribers under it.
+     */
+    @Test
+    fun `FlowSafeSubject wrapped with a FlowSafeSubscriber will preserve the structure, if one of them is unsafe and it throws`() {
```
This comment is confusing to me. From what I understand, the test is checking the opposite of what the comment is saying.

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+            throw IllegalStateException()
+        },/*onError*/{
+            heartBeat++
+        })
```
Maybe write this like this:

```kotlin
source.subscribe({ onNext ->
            heartBeat++
            throw IllegalStateException()
        },{ onError ->
            heartBeat++
        })
```

kyriathar replied:
Just removed comments because, since `heartBeat `was splitted into `heartBeatOnNext` and `heartBeatOnError ` blocks are now well implied

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+
+        source.onNext(1)
+        source.onNext(1)
+        assertThat(heartBeat == 4)
```
Add separate counts for `onNext` and `onError` (or even just test `onError` counts here)

kyriathar replied:
Splitted `heartBeat `into `heartBeatOnNext` and `heartBeatOnError`

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+    }
+
+    @Test
+    fun `throwing FlowSafeSubscriber as a leaf will call onError`() {
```
```
throwing FlowSafeSubscriber as a leaf will call onError
```

to

```
throwing during onNext of a FlowSafeSubscriber leaf subscriber will call onError
```

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+    }
+
+    @Test
+    fun `throwing FlowSafeSubscriber at onNext will wrap with a Rx OnErrorNotImplementedException`() {
```
Does this test depend on the `FlowSafeSubscriber` being a lead subscriber? In this case, we should say so.

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+    }
+
+    @Test
+    fun `throwing FlowSafeSubscriber at onError will wrap with a Rx OnErrorFailedException`() {
```
`throwing inside FlowSafeSubscriber.onError will wrap with a Rx OnErrorFailedException`

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -186,6 +199,171 @@ class ObservablesTests {
+        source.subscribe { throw IllegalStateException("123") } // will give a leaf FlowSafeSubscriber
+        val sourceWrapper = FlowSafeSubscriber(Subscribers.from(source)) // will give an inner FlowSafeSubscriber
+
+        assertFailsWith<OnErrorNotImplementedException>("123") {
```
Is this meant to be `OnErrorNotImplementedException` or `OnNextFailedException`? The issue using `assertFailsWith` in this scenario is that it does not distinguish between them (as `OnNextFailedException` is a subclass of `OnErrorNotImplementedException`).

kyriathar replied:
Valid point, It is a `OnNextFailedException`, will update test

lankydan commented on testing/cordapps/dbfailure/dbfworkflows/src/main/kotlin/com/r3/dbfailure/workflows/SendStateFlow.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    @StartableByRPC
+    class NotarisedTxs : FlowLogic<List<String>>() {
```
Can we keep this in the test class as it is not really got anything to do with the testing cordapp.

lankydan commented on testing/cordapps/dbfailure/dbfworkflows/src/main/kotlin/com/r3/transactionfailure/workflows/ErrorHandling.kt:
```diff
@@ -51,4 +52,15 @@ object ErrorHandling {
     }
 
+    @StartableByRPC
+    class SubscribingRawUpdatesFlow: FlowLogic<Unit>() {
```
Add this into the test class instead of the cordapp.

lankydan replied:
I don't think you committed the test class for this either.

kyriathar replied:
test method for this class is: `VaultObserverExceptionTest.Subscribing to NodeVaultService rawUpdates from a flow is not allowed` - will move `SubscribingRawUpdatesFlow `under test  class

lankydan commented:
Also realized that I don't see the `flowSafeSubscribe` function anywhere. Was it not committed?

kyriathar-r3 committed:
Add kdoc to OnNextFailedException

kyriathar-r3 committed:
Minor text correction

kyriathar-r3 committed:
Update kdocs of FlowSafeSubject/ PreventSubscriptionsSubject

kyriathar-r3 committed:
Moved FlowSafeSubject under package node.internal as it is only used by NodeVaultService

kyriathar-r3 committed:
Add comment and update kdoc explaining how to subscribe with SafeSubscriber to FlowSafeSubject

kyriathar-r3 committed:
Change PreventSubscriptionsSubject#errorAction to be more specific; to return an Exception

kyriathar-r3 committed:
Minor text update

kyriathar-r3 committed:
Update messy comment

kyriathar-r3 committed:
Replace assertThat with assertEquals

kyriathar-r3 committed:
Splitting heartBeat to heartBeat1 and hearBeat2 for more clear asserting

kyriathar-r3 committed:
Correcting comment

kyriathar-r3 committed:
Update messy comment

lankydan commented on node/src/main/kotlin/net/corda/node/internal/FlowSafeSubject.kt:
```diff
@@ -0,0 +1,51 @@
+/**
+ * [FlowSafeSubject] is used to unwrap an [Observer] from a [SafeSubscriber], re-wrap it with a [FlowSafeSubscriber]
+ * and then subscribe it to its underlying [Subject]. It is only used to provide therefore, its underlying [Subject] with
+ * non unsubscribing [rx.Observer]s.
```
I dont understand this sentence.

```
It is only used to provide therefore, its underlying [Subject] with non unsubscribing [rx.Observer]s.
```

lankydan commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -209,7 +212,31 @@ class NodeVaultService(
+                // because the Observer could reference flow's properties, essentially fiber's properties then,
+                // since it does not unsubscribe on flow's/ fiber's completion,
+                // it could prevent the flow/ fiber -object- get garbage collected.
+                return PreventSubscriptionsSubject(_rawUpdatesPublisher) {
```
> Yes good point, i just thought because it is public API, then make least possible changes. In that sense, still return the rx.Observable but prevent from subscribing to it.

I think just straight throwing an exception is fine here. The only reason to access `rawUpdates` is to do some sort of subscribing or other operation that depends on events coming from it. There are probably a few other things, but I do not see any reason why they would need a good reason to access it.

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -214,6 +223,150 @@ class ObservablesTests {
+            source.onNext(2) // first observer will run, second observer will run and throw
+        }
+        source.onNext(3) // both observers will run
+        assertEquals(3, heartBeat1)
```
> Is it correct that the heartBeat is 5? Shouldn't it increase by 2 for each event passed into onNext? Making it 6?
> 
> I think it would also be worth splitting out the heartBeat counts into multiple variables. So it is 100% clear how many times each subscriber is called (since there is only 2 of them)
> 
> i think 5 is correct;
> run_1: the first Observer will increase counter and throw - the second will not be executed (+1)
> run_2: first will increase counter - second will increase counter and throw (+2)
> run_3: both Observers will run normally (+2)
> 
> heartBeat splitted into 2 variables

Ah so that is because the 2nd subscriber's `onNext` doesn't trigger due to the first subscriber's exception?

kyriathar-r3 committed:
Splitting heartBeat into heartBeatOnNext and heartBeatOnError

kyriathar-r3 committed:
Update test name

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -214,6 +223,150 @@ class ObservablesTests {
+     * A FlowSafeSubscriber that is NOT a leaf in a Subscribers structure, if it throws at onNext, it will not call its onError.
+     *
+     *  The reason why it should not call its onError is: if it wraps a [PublishSubject], calling [FlowSafeSubscriber.onError]
+     *  will then call [PublishSubject.onError] which will shut down all the Subscribers under the [PublishSubject].
```
The second part of the comment is indented too much.

The first sentence would make more sense rewritten to:

```
A [FlowSafeSubscriber] that is NOT a leaf in a Subscribers structure will not call [onError] if an error occurs during its [onNext] event.
```

kyriathar-r3 committed:
Add explanatory comment to test

kyriathar-r3 committed:
Update test name

kyriathar-r3 committed:
Update test and add test comment

kyriathar-r3 committed:
Moving NotarisedTxs from SendStateFlow to VaultObserverExceptionTest inside NodeHandle.getNotarisedTransactionIds

kyriathar-r3 committed:
Moving SubscribingRawUpdatesFlow from ErrorHandling to VaultObserverExceptionTest

kyriathar-r3 committed:
Update kdoc of FlowSafeSubscriber and FlowSafeSubscriber.onNext

kyriathar-r3 committed:
Make kdoc more clear

kyriathar-r3 committed:
Throw exception upon accessing VaultService.rawUpdates from within a flow

kyriathar-r3 committed:
Changing exception thrown when accessing VaultService.rawUpdates from within a flow to a CordaRuntimeException

kyriathar-r3 committed:
Minor kdoc update

kyriathar-r3 committed:
Update test comment

kyriathar-r3 committed:
Update kdoc of FlowSafeSubscriber

lankydan commented on node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt:
```diff
@@ -209,7 +211,27 @@ class NodeVaultService(
+
+                throw CordaRuntimeException(
+                    "Flow ${it.logic::class.java.name} tried to access VaultService.rawUpdates " +
+                            "- Rx.Observables should only be accessed to outside the context of a flow "
```
`- Rx.Observables should only be accessed outside the context of a flow`

lankydan commented on node/src/test/kotlin/net/corda/node/utilities/ObservablesTests.kt:
```diff
@@ -214,6 +223,165 @@ class ObservablesTests {
         assertEquals(2, count)
     }
 
+    @Test
```
We need to remember to add `@Test(timeout=300_000)` from now on to tests (or some suitable timeout. You'll need to replace these on the tests.

kyriathar-r3 committed:
Introducing Observable.flowSafeSubscribe public API method to subscribe with -non unsubscribing- Rx.Subscribers to Observables. It also replaced FlowSafeSubject

kyriathar-r3 committed:
Move CustomSafeSubscriber outside test methods

kyriathar-r3 committed:
Minor text update

kyriathar-r3 committed:
Add timeout to tests

kyriathar-r3 committed:
Update kdoc of flowSafeSubscribe

kyriathar-r3 committed:
Update kdoc of flowSafeSubscribe

kyriathar-r3 committed:
Update kdoc of flowSafeSubscribe

kyriathar-r3 committed:
Move FlowSafeSubscriber and flowSafeSubscribe under their own package

kyriathar-r3 committed:
Fix detekt issue

kyriathar-r3 committed:
Update Detekt baseline

kyriathar-r3 committed:
Revert "Update Detekt baseline"

This reverts commit 793a8ed9

kyriathar-r3 committed:
Fix Detekt issue

kyriathar-r3 committed:
Moved strictMode flag from flowSafeSubscribe to OnFlowSafeSubscribe

Moved OnFlowSafeSubscribe into internal package

Integration tested flowSafeLooseSubscribe

kyriathar-r3 committed:
Suppress Rx Deprecation

kyriathar-r3 committed:
Rename flowSafeSubscribe to flowSafeObservable

kyriathar-r3 committed:
Renaming flowSafeObservable to continueOnError and FlowSafeSubscriber to ResilientSubscriber

lankydan commented on core/src/main/kotlin/net/corda/core/observable/internal/ResilientSubscriber.kt:
```diff
@@ -0,0 +1,120 @@
+
+/**
+ * [OnResilientSubscribe] returns an [Observable] holding a reference to the source [Observable]. Upon subscribing to it,
+ * when reaching [call] method, if the subscriber passed in [isSafeSubscriber] it will unwrap the [Observer] from
```
`[isSafeSubscriber]` to `is a [SafeSubscriber]`

lankydan reviewed (approved):
Left 1 comment but we can do that in a separate PR. We will need another PR to add documentation around this area. But for now, I want to get this merged in.
